### PR TITLE
Rework srpms output set creation

### PR DESCRIPTION
**Pull request by rbikar**
Reworked srpms output set creation. Now it doesn't guess filename of srpm if its filename is not referenced on binary rpm unit.
Updated tests.

### CODE REVIEW

File: ubipop/__init__.py [hunk 1]
```diff
@@ -271,24 +270,23 @@ def _finalize_output_units(self, units, type_id):
...
             if package.sourcerpm_filename is None:
-                name, ver, rel, _, _ = splitFilename(package.filename)
-                package.sourcerpm_filename = "{n}-{v}-{r}.src.rpm".format(n=name, v=ver, r=rel)
+                _LOG.warning("Package %s doesn't reference its source rpm", package.name)
+                continue
 
-            self.repos.source_rpms.append(Package(package.name,
-                                                  package.sourcerpm_filename))
+            self.repos.source_rpms[package.name].extend([Package(package.name,
+                                                                 package.sourcerpm_filename)])
```
Comment (JayZ12138) [inline ubipop/__init__.py]: .extend([Package(package.name,
                                                  package.sourcerpm_filename))	                                                                 package.sourcerpm_filename)])

I don't understand this change, seems completely unnecessary, can you explain?

File: ubipop/__init__.py [hunk 2]
```diff
@@ -271,24 +270,23 @@ def _finalize_output_units(self, units, type_id):
             self.keep_n_latest_modules(units)
 
     def _create_srpms_output_set(self):
-        packages = chain.from_iterable(self.repos.packages.values())
-        for package in packages:
+        rpms = list(chain.from_iterable(self.repos.packages.values()))
```
Comment (JayZ12138) [inline ubipop/__init__.py]: why need to make it a list here?

File: ubipop/__init__.py (hunk 1 above)
Comment (JayZ12138) [reply ubipop/__init__.py]: I mean, append the package instead of making it a list and then extend should work properly.

Comment (rbikar) [reply ubipop/__init__.py]: Yeah, appending just Package is sufficient here.

File: ubipop/__init__.py (hunk 2 above)
Comment (rbikar) [reply ubipop/__init__.py]: Yes, you're right. It's not needed here. It iterates only once over rpms, so it will ok, to leave a chain obj.

Comment (alexandrevicenzi) [reply ubipop/__init__.py]: My question is, why do you create a generator from a list to cast to a list again? Could just be:

`rpms = self.repos.packages.values()`

Right?

File: ubipop/__init__.py (hunk 1 above)
Comment (alexandrevicenzi) [reply ubipop/__init__.py]: `extend` is more time consuming than `append`.

File: ubipop/__init__.py [hunk 3]
```diff
@@ -271,24 +270,23 @@ def _finalize_output_units(self, units, type_id):
...
-            self.repos.source_rpms.append(Package(package.name,
-                                                  package.sourcerpm_filename))
+            self.repos.source_rpms[package.name].extend([Package(package.name,
+                                                                 package.sourcerpm_filename)])
 
-        blacklisted = self.get_blacklisted_packages(self.repos.source_rpms)
-        self.repos.source_rpms = \
-            self._diff_packages_by_filename(self.repos.source_rpms, blacklisted)
+        blacklisted_srpms = self.get_blacklisted_packages(
+            list(chain.from_iterable(self.repos.source_rpms.values())))
```
Comment (alexandrevicenzi) [inline ubipop/__init__.py]: Same as before, is this really needed?

File: ubipop/__init__.py [hunk 4]
```diff
@@ -271,24 +270,23 @@ def _finalize_output_units(self, units, type_id):
...
+
+        for pkg in blacklisted_srpms:
+            self.repos.source_rpms.pop(pkg.name, None)
 
     def _determine_pulp_actions(self, units, current, diff_f):
-        if isinstance(units, dict):
-            expected = list(chain.from_iterable(units.values()))
-        else:
-            expected = units
+        expected = list(chain.from_iterable(units.values()))
```
Comment (alexandrevicenzi) [inline ubipop/__init__.py]: Same as before, is this really needed?

File: ubipop/__init__.py (hunk 2 above)
Comment (rbikar) [reply ubipop/__init__.py]: self.repos.packages is {"key": ["v1", "v2]} so self.repos.packages.values() is list of lists and I need to flatten it.

File: ubipop/__init__.py (hunk 4 above)
Comment (rbikar) [reply ubipop/__init__.py]: Yes, it iterates several times over expected further in code, so generator is not enough.

File: ubipop/__init__.py (hunk 3 above)
Comment (rbikar) [reply ubipop/__init__.py]: Yes, in self.get_blacklisted_packages() it iterates several times over blacklisted_srpms, so generator is not enough.

**Pull request merged**